File: clients/lumify-sdk/src/resources/players.ts
import type { LumifyClient } from "../client.js";
import type { Player, PlayerEventsResponse, PlayersListResponse } from "../generated/models.js";
import { paginate, iterateItems, type PaginateOptions } from "../pagination.js";

export interface ListPlayersParams {
  sport?: string;
  /** Free-text search on player name. */
  q?: string;
  country?: string;
  active?: boolean;
  /** Only players with a ranking (e.g. tennis ATP/WTA). */
  ranked?: boolean;
  afterId?: number;
  /** Max 100, default 25. */
  limit?: number;
}

export interface PlayerEventsParams {
  status?: string;
  from?: string;
  to?: string;
  afterId?: number;
  limit?: number;
}

function toQuery(p: ListPlayersParams) {
  return {
    sport: p.sport,
    q: p.q,
    country: p.country,
    active: p.active,
    ranked: p.ranked,
    after_id: p.afterId,
    limit: p.limit,
  };
}

export class PlayersResource {
  constructor(private readonly client: LumifyClient) {}

  /** GET /v1/players — cursor-paginated. */
  list(params: ListPlayersParams = {}): Promise<PlayersListResponse> {
    return this.client.get<PlayersListResponse>("/v1/players", { query: toQuery(params) });
  }

  /** Async-iterate every page of `list()` for the given filters. */
  paginate(params: Omit<ListPlayersParams, "afterId"> = {}, options: PaginateOptions = {}) {
    return paginate(
      (afterId, limit) => this.list({ ...params, afterId, limit }),
      { limit: params.limit, ...options }
    );
  }

  /** Async-iterate every player matching the filters, across all pages. */
  iterate(params: Omit<ListPlayersParams, "afterId"> = {}, options: PaginateOptions = {}) {
    return iterateItems(
      (afterId, limit) => this.list({ ...params, afterId, limit }),
      { limit: params.limit, ...options }
    );
  }

  /** GET /v1/players/{id} */
  get(playerId: number): Promise<Player> {
    return this.client.get<Player>(`/v1/players/${playerId}`);
  }

  /** GET /v1/players/{id}/events — cursor-paginated schedule/results for a player. */
  events(playerId: number, params: PlayerEventsParams = {}): Promise<PlayerEventsResponse> {
    return this.client.get<PlayerEventsResponse>(`/v1/players/${playerId}/events`, {
      query: {
        status: params.status,
        from: params.from,
        to: params.to,
        after_id: params.afterId,
        limit: params.limit,
      },
    });
  }

  /** Async-iterate every page of `events()` for the given player/filters. */
  paginateEvents(playerId: number, params: Omit<PlayerEventsParams, "afterId"> = {}, options: PaginateOptions = {}) {
    return paginate(
      (afterId, limit) => this.events(playerId, { ...params, afterId, limit }),
      { limit: params.limit, ...options }
    );
  }
}

File: clients/lumify-sdk-python/lumify/resources/players.py
from __future__ import annotations

from typing import Any, Iterator, Optional

from .._transport import LumifyClient
from ..models import Player, PlayerEventsResponse, PlayersListResponse
from ..pagination import iterate_items, paginate


class PlayersResource:
    def __init__(self, client: LumifyClient) -> None:
        self._client = client

    def list(
        self,
        *,
        sport: Optional[str] = None,
        q: Optional[str] = None,
        country: Optional[str] = None,
        active: Optional[bool] = None,
        ranked: Optional[bool] = None,
        after_id: Optional[int] = None,
        limit: Optional[int] = None,
    ) -> PlayersListResponse:
        """GET /v1/players — cursor-paginated."""
        return self._client.get(
            "/v1/players",
            query={
                "sport": sport,
                "q": q,
                "country": country,
                "active": active,
                "ranked": ranked,
                "after_id": after_id,
                "limit": limit,
            },
        )

    def paginate(
        self, *, limit: int = 25, max_pages: int = 1000, **filters: Any
    ) -> Iterator[PlayersListResponse]:
        return paginate(
            lambda after_id, lim: self.list(after_id=after_id, limit=lim, **filters),
            limit=limit,
            max_pages=max_pages,
        )

    def iterate(
        self, *, limit: int = 25, max_pages: int = 1000, **filters: Any
    ) -> Iterator[Any]:
        return iterate_items(
            lambda after_id, lim: self.list(after_id=after_id, limit=lim, **filters),
            items_key="data",
            limit=limit,
            max_pages=max_pages,
        )

    def get(self, player_id: int) -> Player:
        """GET /v1/players/{id}."""
        return self._client.get("/v1/players/%d" % player_id)

    def events(
        self,
        player_id: int,
        *,
        status: Optional[str] = None,
        from_: Optional[str] = None,
        to: Optional[str] = None,
        after_id: Optional[int] = None,
        limit: Optional[int] = None,
    ) -> PlayerEventsResponse:
        """GET /v1/players/{id}/events — cursor-paginated schedule/results.
        ``from_`` maps to the ``from`` query param (a Python keyword)."""
        return self._client.get(
            "/v1/players/%d/events" % player_id,
            query={
                "status": status,
                "from": from_,
                "to": to,
                "after_id": after_id,
                "limit": limit,
            },
        )

    def paginate_events(
        self, player_id: int, *, limit: int = 25, max_pages: int = 1000, **filters: Any
    ) -> Iterator[PlayerEventsResponse]:
        return paginate(
            lambda after_id, lim: self.events(
                player_id, after_id=after_id, limit=lim, **filters
            ),
            limit=limit,
            max_pages=max_pages,
        )
